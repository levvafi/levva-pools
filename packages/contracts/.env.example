ETH_NODE_URL=
ARB_NODE_URL=

ETH_API_KEY=
ARB_API_KEY=