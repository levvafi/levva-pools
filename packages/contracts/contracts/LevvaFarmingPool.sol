// SPDX-License-Identifier: BUSL-1.1
pragma solidity 0.8.28;

import './base/trading/LongTrading.sol';
import './base/farming/ShortFarming.sol';
import './base/Emergency.sol';

contract LevvaFarmingPool is LongTrading, ShortFarming, Emergency {
  using FP96 for FP96.FixedPoint;
  using LowGasSafeMath for uint256;
  using MaxBinaryHeapLib for MaxBinaryHeapLib.Heap;

  /// @inheritdoc IMarginlyPool
  function initialize(
    address _quoteToken,
    address _baseToken,
    address _priceOracle,
    uint32 _defaultSwapCallData,
    MarginlyParams calldata _params
  ) external virtual {
    if (factory != address(0)) revert MarginlyErrors.Forbidden();

    __LevvaPoolCommon_init(_quoteToken, _baseToken, _priceOracle, _defaultSwapCallData, _params);
    __LongTrading_init();
  }

  /// @param flag unwrapETH in case of withdraw calls or syncBalance in case of reinit call
  function execute(
    CallType call,
    uint256 amount1,
    int256 amount2,
    uint256 limitPriceX96,
    bool flag,
    address positionAddress,
    uint256 swapCalldata
  ) external payable override lock {
    if (call == CallType.ReceivePosition) {
      if (amount2 < 0) revert MarginlyErrors.WrongValue();
      _receivePosition(positionAddress, amount1, uint256(amount2));
      return;
    } else if (call == CallType.EmergencyWithdraw) {
      _emergencyWithdraw(flag);
      return;
    }

    if (mode != Mode.Regular) revert MarginlyErrors.EmergencyMode();

    (bool callerMarginCalled, FP96.FixedPoint memory basePrice) = _reinit();
    if (callerMarginCalled) {
      _updateSystemLeverages(basePrice);
      return;
    }

    (Position storage position, address positionOwner) = _getPosition(positionAddress);

    if (_positionHasBadLeverage(position, basePrice)) {
      _liquidate(msg.sender, position, basePrice);
      _updateSystemLeverages(basePrice);
      return;
    }

    if (call == CallType.DepositBase) {
      _depositBase(amount1, basePrice, position, positionOwner);
      if (amount2 > 0) {
        _long(uint256(amount2), limitPriceX96, basePrice, position, positionOwner, swapCalldata);
      } else if (amount2 < 0) {
        _short(uint256(-amount2), limitPriceX96, basePrice, position, positionOwner, swapCalldata);
      }
    } else if (call == CallType.DepositQuote) {
      _depositQuote(amount1, position, positionOwner);
      if (amount2 > 0) {
        _short(uint256(amount2), limitPriceX96, basePrice, position, positionOwner, swapCalldata);
      } else if (amount2 < 0) {
        _long(uint256(-amount2), limitPriceX96, basePrice, position, positionOwner, swapCalldata);
      }
    } else if (call == CallType.WithdrawBase) {
      _withdrawBase(amount1, flag, basePrice, position);
    } else if (call == CallType.WithdrawQuote) {
      _withdrawQuote(amount1, flag, basePrice, position);
    } else if (call == CallType.Short) {
      _short(amount1, limitPriceX96, basePrice, position, positionOwner, swapCalldata);
    } else if (call == CallType.Long) {
      _long(amount1, limitPriceX96, basePrice, position, positionOwner, swapCalldata);
    } else if (call == CallType.ClosePosition) {
      _closePosition(limitPriceX96, position, swapCalldata);
    } else if (call == CallType.SellCollateral) {
      _sellCollateral(limitPriceX96, position, swapCalldata);
      if (flag) {
        _withdrawQuote(type(uint256).max, false, basePrice, position);
      }
    } else if (call == CallType.Reinit && flag) {
      // reinit itself has already taken place
      _syncBaseBalance();
      _syncQuoteBalance();
      emit BalanceSync();
    }

    _updateHeap(position);
    _updateSystemLeverages(basePrice);
  }

  /// @dev Used by keeper service
  function getHeapPosition(
    uint32 index,
    bool _short
  ) external view returns (bool success, MaxBinaryHeapLib.Node memory node) {
    if (_short) {
      return (false, node);
    } else {
      return longHeap.getNodeByIndex(index);
    }
  }

  function _calcRealBaseCollateralTotal() internal view override(LevvaPoolCommon, LongTrading) returns (uint256) {
    return LongTrading._calcRealBaseCollateralTotal();
  }

  function _calcRealQuoteDebtTotal() internal view override(LevvaPoolVirtual, LongTrading) returns (uint256) {
    return LongTrading._calcRealQuoteDebtTotal();
  }

  function _calcRealBaseCollateral(
    uint256 disBaseCollateral,
    uint256 disQuoteDebt
  ) internal view override(LevvaPoolCommon, LongTrading) returns (uint256) {
    return LongTrading._calcRealBaseCollateral(disBaseCollateral, disQuoteDebt);
  }

  function _calcRealQuoteDebt(
    uint256 disQuoteDebt
  ) internal view override(LevvaPoolVirtual, LongTrading) returns (uint256) {
    return LongTrading._calcRealQuoteDebt(disQuoteDebt);
  }

  function _updateBaseCollateralCoeffs(FP96.FixedPoint memory factor) internal override(LevvaPoolCommon, LongTrading) {
    return LongTrading._updateBaseCollateralCoeffs(factor);
  }

  function _deleverageLong(
    uint256 realQuoteCollateral,
    uint256 realBaseDebt
  ) internal override(LongTrading, LevvaPoolVirtual) {
    LongTrading._deleverageLong(realQuoteCollateral, realBaseDebt);
  }

  function _getWorstLongPositionOwner() internal view override(LongTrading, LevvaPoolVirtual) returns (address) {
    return LongTrading._getWorstLongPositionOwner();
  }

  function _updateHeapLong(Position storage position) internal override(LongTrading, LevvaPoolVirtual) {
    return LongTrading._updateHeapLong(position);
  }
}
