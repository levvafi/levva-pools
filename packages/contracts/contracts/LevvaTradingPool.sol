// SPDX-License-Identifier: BUSL-1.1
pragma solidity 0.8.28;

import './pool/base/trading/LongTrading.sol';
import './pool/base/trading/ShortTrading.sol';
import './pool/base/Emergency.sol';

contract LevvaTradingPool is LongTrading, ShortTrading, Emergency {
  using FP96 for FP96.FixedPoint;
  using LowGasSafeMath for uint256;
  using MaxBinaryHeapLib for MaxBinaryHeapLib.Heap;

  /// @inheritdoc IMarginlyPool
  function initialize(
    address _quoteToken,
    address _baseToken,
    address _priceOracle,
    uint32 _defaultSwapCallData,
    MarginlyParams calldata _params
  ) external virtual {
    if (factory != address(0)) revert MarginlyErrors.Forbidden();

    __LevvaPoolCommon_init(_quoteToken, _baseToken, _priceOracle, _defaultSwapCallData, _params);
    __LongTrading_init();
    __ShortTrading_init();
  }

  /// @param flag unwrapETH in case of withdraw calls or syncBalance in case of reinit call
  function execute(
    CallType call,
    uint256 amount1,
    int256 amount2,
    uint256 limitPriceX96,
    bool flag,
    address positionAddress,
    uint256 swapCalldata
  ) external payable override lock {
    if (call == CallType.ReceivePosition) {
      if (amount2 < 0) revert MarginlyErrors.WrongValue();
      _receivePosition(positionAddress, amount1, uint256(amount2));
      return;
    } else if (call == CallType.EmergencyWithdraw) {
      _emergencyWithdraw(flag);
      return;
    }

    if (mode != Mode.Regular) revert MarginlyErrors.EmergencyMode();

    (bool callerMarginCalled, FP96.FixedPoint memory basePrice) = _reinit();
    if (callerMarginCalled) {
      _updateSystemLeverages(basePrice);
      return;
    }

    (Position storage position, address positionOwner) = _resolvePositionAndOwner(positionAddress);

    if (_positionHasBadLeverage(position, basePrice)) {
      _liquidate(msg.sender, position, basePrice);
      _updateSystemLeverages(basePrice);
      return;
    }

    if (call == CallType.DepositBase) {
      _depositBase(amount1, basePrice, position, positionOwner);
      if (amount2 > 0) {
        _long(uint256(amount2), flag, limitPriceX96, basePrice, position, positionOwner, swapCalldata);
      } else if (amount2 < 0) {
        _short(uint256(-amount2), flag, limitPriceX96, basePrice, position, positionOwner, swapCalldata);
      }
    } else if (call == CallType.DepositQuote) {
      _depositQuote(amount1, position, positionOwner);
      if (amount2 > 0) {
        _short(uint256(amount2), flag, limitPriceX96, basePrice, position, positionOwner, swapCalldata);
      } else if (amount2 < 0) {
        _long(uint256(-amount2), flag, limitPriceX96, basePrice, position, positionOwner, swapCalldata);
      }
    } else if (call == CallType.WithdrawBase) {
      _withdrawBase(amount1, flag, basePrice, position);
    } else if (call == CallType.WithdrawQuote) {
      _withdrawQuote(amount1, flag, basePrice, position);
    } else if (call == CallType.Short) {
      _short(amount1, flag, limitPriceX96, basePrice, position, positionOwner, swapCalldata);
    } else if (call == CallType.Long) {
      _long(amount1, flag, limitPriceX96, basePrice, position, positionOwner, swapCalldata);
    } else if (call == CallType.ClosePosition) {
      _closePosition(limitPriceX96, position, swapCalldata);
    } else if (call == CallType.SellCollateral) {
      _sellCollateral(limitPriceX96, position, positionOwner, swapCalldata);
    } else if (call == CallType.Reinit && flag) {
      // reinit itself has already taken place
      _syncBaseBalance();
      _syncQuoteBalance();
      emit BalanceSync();
    }

    _updateHeap(position);
    _updateSystemLeverages(basePrice);
  }

  /// @dev Used by keeper service
  function getHeapPosition(
    uint32 index,
    bool _short
  ) external view returns (bool success, MaxBinaryHeapLib.Node memory) {
    if (_short) {
      return shortHeap.getNodeByIndex(index);
    } else {
      return longHeap.getNodeByIndex(index);
    }
  }

  function _calcRealBaseCollateralTotal() internal view override(LevvaPoolCommon, LongTrading) returns (uint256) {
    return LongTrading._calcRealBaseCollateralTotal();
  }

  function _calcRealQuoteCollateralTotal() internal view override(LevvaPoolCommon, ShortTrading) returns (uint256) {
    return ShortTrading._calcRealQuoteCollateralTotal();
  }

  function _calcRealBaseDebtTotal() internal view override(LevvaPoolVirtual, ShortTrading) returns (uint256) {
    return ShortTrading._calcRealBaseDebtTotal();
  }

  function _calcRealQuoteDebtTotal() internal view override(LevvaPoolVirtual, LongTrading) returns (uint256) {
    return LongTrading._calcRealQuoteDebtTotal();
  }

  function _calcRealBaseCollateral(
    uint256 disBaseCollateral,
    uint256 disQuoteDebt
  ) internal view override(LevvaPoolCommon, LongTrading) returns (uint256) {
    return LongTrading._calcRealBaseCollateral(disBaseCollateral, disQuoteDebt);
  }

  function _calcRealQuoteCollateral(
    uint256 disQuoteCollateral,
    uint256 disBaseDebt
  ) internal view override(LevvaPoolCommon, ShortTrading) returns (uint256) {
    return ShortTrading._calcRealQuoteCollateral(disQuoteCollateral, disBaseDebt);
  }

  function _calcRealBaseDebt(
    uint256 disBaseDebt
  ) internal view override(LevvaPoolVirtual, ShortTrading) returns (uint256) {
    return ShortTrading._calcRealBaseDebt(disBaseDebt);
  }

  function _calcRealQuoteDebt(
    uint256 disQuoteDebt
  ) internal view override(LevvaPoolVirtual, LongTrading) returns (uint256) {
    return LongTrading._calcRealQuoteDebt(disQuoteDebt);
  }

  function _updateBaseCollateralCoeffs(FP96.FixedPoint memory factor) internal override(LevvaPoolCommon, LongTrading) {
    return LongTrading._updateBaseCollateralCoeffs(factor);
  }

  function _updateQuoteCollateralCoeffs(
    FP96.FixedPoint memory factor
  ) internal override(LevvaPoolCommon, ShortTrading) {
    return ShortTrading._updateQuoteCollateralCoeffs(factor);
  }

  function _deleverageLong(
    uint256 realQuoteCollateral,
    uint256 realBaseDebt
  ) internal override(LongTrading, LevvaPoolVirtual) {
    LongTrading._deleverageLong(realQuoteCollateral, realBaseDebt);
  }

  function _deleverageShort(
    uint256 realBaseCollateral,
    uint256 realQuoteDebt
  ) internal override(ShortTrading, LevvaPoolVirtual) {
    ShortTrading._deleverageShort(realBaseCollateral, realQuoteDebt);
  }

  function _getWorstLongPositionOwner() internal view override(LongTrading, LevvaPoolVirtual) returns (address) {
    return LongTrading._getWorstLongPositionOwner();
  }

  function _getWorstShortPositionOwner() internal view override(ShortTrading, LevvaPoolVirtual) returns (address) {
    return ShortTrading._getWorstShortPositionOwner();
  }

  function _updateHeapLong(Position storage position) internal override(LongTrading, LevvaPoolVirtual) {
    return LongTrading._updateHeapLong(position);
  }

  function _updateHeapShort(Position storage position) internal override(ShortTrading, LevvaPoolVirtual) {
    return ShortTrading._updateHeapShort(position);
  }
}
