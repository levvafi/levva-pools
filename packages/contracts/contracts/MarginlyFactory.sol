// SPDX-License-Identifier: BUSL-1.1
pragma solidity 0.8.28;

import '@uniswap/v3-core/contracts/interfaces/IUniswapV3Factory.sol';
import '@openzeppelin/contracts/proxy/Clones.sol';
import '@openzeppelin/contracts/access/Ownable2Step.sol';

import './pool/interfaces/IMarginlyFactory.sol';
import './pool/interfaces/IMarginlyPool.sol';
import './pool/dataTypes/MarginlyParams.sol';
import './pool/libraries/MarginlyErrors.sol';

/// @title Marginly contract factory
/// @notice Deploys Marginly and manages ownership and control over pool
contract MarginlyFactory is IMarginlyFactory, Ownable2Step {
  address public immutable marginlyPoolImplementation;
  /// @notice Address of uniswap swap router
  address public override swapRouter;
  /// @notice Swap fee holder
  address public immutable override feeHolder;
  /// @notice Address of wrapped ETH
  address public immutable override WETH9;
  /// @notice Technical position address
  address public immutable override techPositionOwner;

  constructor(
    address _marginlyPoolImplementation,
    address _swapRouter,
    address _feeHolder,
    address _WETH9,
    address _techPositionOwner
  ) Ownable(msg.sender) {
    if (
      _marginlyPoolImplementation == address(0) ||
      _swapRouter == address(0) ||
      _feeHolder == address(0) ||
      _WETH9 == address(0) ||
      _techPositionOwner == address(0)
    ) revert MarginlyErrors.WrongValue();

    marginlyPoolImplementation = _marginlyPoolImplementation;
    swapRouter = _swapRouter;
    feeHolder = _feeHolder;
    WETH9 = _WETH9;
    techPositionOwner = _techPositionOwner;
  }

  /// @inheritdoc IMarginlyFactory
  function createPool(
    address quoteToken,
    address baseToken,
    address priceOracle,
    uint32 defaultSwapCallData,
    MarginlyParams calldata params
  ) external override onlyOwner returns (address pool) {
    if (quoteToken == baseToken) revert MarginlyErrors.Forbidden();
    if (priceOracle == address(0)) revert MarginlyErrors.WrongValue();

    pool = Clones.clone(marginlyPoolImplementation);
    IMarginlyPool(pool).initialize(quoteToken, baseToken, priceOracle, defaultSwapCallData, params);

    emit PoolCreated(quoteToken, baseToken, priceOracle, defaultSwapCallData, pool);
  }

  /// @inheritdoc IMarginlyFactory
  function changeSwapRouter(address newSwapRouter) external onlyOwner {
    if (newSwapRouter == address(0)) revert MarginlyErrors.WrongValue();
    swapRouter = newSwapRouter;
    emit SwapRouterChanged(newSwapRouter);
  }

  function renounceOwnership() public view override onlyOwner {
    revert MarginlyErrors.Forbidden();
  }
}
