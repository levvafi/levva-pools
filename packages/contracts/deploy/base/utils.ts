import { ethers } from 'ethers';

export function validateAddress(address: string) {
  if (!ethers.isAddress(address) || address === ethers.ZeroAddress) {
    throw new Error(`Wrong ${address} evm address`);
  }
}
