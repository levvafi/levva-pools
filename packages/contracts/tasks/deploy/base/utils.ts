import { ethers } from 'ethers';

export function validateAddress(address: string) {
  if (!ethers.isAddress(address) || address === ethers.ZeroAddress) {
    throw new Error(`Wrong ${address} evm address`);
  }
}

export function isSameAddress(addressA: string, addressB: string): boolean {
  return ethers.getAddress(addressA) === ethers.getAddress(addressB);
}
