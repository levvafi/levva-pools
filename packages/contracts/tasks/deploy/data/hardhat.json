{
  "factory": {
    "poolType": "Trading",
    "feeHolderAddress": "0xd6DE5D2D8Ab2E4a92c2B923a88f7D366ddb3C760",
    "techPositionOwnerAddress": "0xd6DE5D2D8Ab2E4a92c2B923a88f7D366ddb3C760",
    "WETH9": {
      "address": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
      "decimals": 18,
      "symbol": "WETH"
    }
  },
  "pools": [
    {
      "priceOracle": "0xd5Ea994A033737f0D568A2A307c65f8048b10698",
      "baseToken": {
        "address": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
        "decimals": 18,
        "symbol": "WETH"
      },
      "quoteToken": {
        "address": "0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48",
        "decimals": 6,
        "symbol": "USDC"
      },
      "defaultSwapCallData": 0,
      "params": {
        "maxLeverage": 20,
        "interestRate": 50000,
        "fee": 20000,
        "swapFee": 500,
        "mcSlippage": 50000,
        "positionMinAmount": 1e17,
        "quoteLimit": 100000e6
      }
    }
  ],
  "oracles": {
    "UniswapV3TickOracle": {
      "factoryAddress": "0x1F98431c8aD98523631AE4a59f267346ea31F984",
      "settings": [
        {
          "quoteToken": {
            "address": "0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48",
            "decimals": 6,
            "symbol": "USDC"
          },
          "baseToken": {
            "address": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
            "decimals": 18,
            "symbol": "WETH"
          },
          "uniswapFee": 500,
          "secondsAgo": 1800,
          "secondsAgoLiquidation": 5
        }
      ]
    }
  },
  "adapters": {
    "UniswapV3Adapter": {
      "dexId": 0,
      "settings": [
        {
          "token0": {
            "address": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
            "decimals": 18,
            "symbol": "WETH"
          },
          "token1": {
            "address": "0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48",
            "decimals": 6,
            "symbol": "USDC"
          },
          "poolAddress": "0x88e6A0c2dDD26FEEb64F039a2c41296FcB3f5640"
        }
      ]
    }
  }
}
