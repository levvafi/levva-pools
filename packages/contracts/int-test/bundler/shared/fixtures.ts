import { SignerWithAddress } from '@nomicfoundation/hardhat-ethers/signers';
import { ethers } from 'hardhat';
import { parseUnits, ZeroAddress } from 'ethers';
import { IERC20__factory, LevvaPendleBundler, LevvaPendleBundler__factory } from '../../../typechain-types';
import { LevvaFarmingPool, LevvaFarmingPool__factory } from '../../../typechain-types';
import { MarginlyParamsStruct } from '../../../typechain-types/contracts/LevvaFarmingPool';
import { MarginlyFactory__factory } from '../../../typechain-types';
import { MarginlyRouter__factory } from '../../../typechain-types';
import { PendleMarketAdapter__factory } from '../../../typechain-types';
import { PendleMarketOracle__factory } from '../../../typechain-types';
import { PENDLE_ROUTER, WETH9, PENDLE_ORACLE, WSTETH } from './registry';
import { PENDLE_PT_WSTETH_30_12_27, PENDLE_MARKET_WSTETH_30_12_27 } from './registry';
import { EthereumMainnetERC20BalanceOfSlot, setTokenBalance } from './tokens';
import { CallType } from '../../pool/utils/chain-ops';

export async function deployPendleBundlerWithPool(): Promise<{ bundler: LevvaPendleBundler; pool: LevvaFarmingPool }> {
  const [signer] = await ethers.getSigners();
  const pool = await deployPool(signer);
  const bundler = await deployPendleBundler(signer);
  return { pool, bundler };
}

async function deployPendleBundler(owner: SignerWithAddress): Promise<LevvaPendleBundler> {
  return new LevvaPendleBundler__factory(owner).deploy(PENDLE_ROUTER);
}

async function deployPool(owner: SignerWithAddress): Promise<LevvaFarmingPool> {
  const poolImplementation = await new LevvaFarmingPool__factory(owner).deploy();

  const pendleAdapter = await new PendleMarketAdapter__factory(owner).deploy([
    {
      pendleMarket: PENDLE_MARKET_WSTETH_30_12_27,
      slippage: 5,
      ptToken: PENDLE_PT_WSTETH_30_12_27,
      ibToken: WSTETH,
    },
  ]);
  const router = await new MarginlyRouter__factory(owner).deploy([
    { dexIndex: 0, adapter: await pendleAdapter.getAddress() },
  ]);

  const oracle = await new PendleMarketOracle__factory(owner).deploy(PENDLE_ORACLE);
  await oracle.setPair(WSTETH, PENDLE_PT_WSTETH_30_12_27, PENDLE_MARKET_WSTETH_30_12_27, 600, 10);

  const factory = await new MarginlyFactory__factory(owner).deploy(poolImplementation, router, owner, WETH9, owner);
  const poolAddress = await factory.createPool.staticCall(
    WSTETH,
    PENDLE_PT_WSTETH_30_12_27,
    oracle,
    0,
    defaultParams()
  );
  await factory.createPool(WSTETH, PENDLE_PT_WSTETH_30_12_27, oracle, 0, defaultParams());

  const pool = LevvaFarmingPool__factory.connect(poolAddress, owner.provider);

  const deposit = parseUnits('50', 18);
  await setTokenBalance(WSTETH, EthereumMainnetERC20BalanceOfSlot.WSTETH, owner.address, deposit);
  await IERC20__factory.connect(WSTETH, owner).approve(pool, deposit);
  await pool.connect(owner).execute(CallType.DepositQuote, deposit, 0, 0, false, ZeroAddress, 0);

  return pool;
}

function defaultParams(): MarginlyParamsStruct {
  return {
    interestRate: 0,
    fee: 0,
    maxLeverage: 20,
    swapFee: 0,
    mcSlippage: 10000,
    positionMinAmount: 1,
    quoteLimit: parseUnits('200', 18),
  };
}
