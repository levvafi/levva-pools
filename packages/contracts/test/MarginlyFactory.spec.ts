import { expect } from 'chai';
import { loadFixture } from '@nomicfoundation/hardhat-network-helpers';
import { MarginlyParamsStruct } from '../typechain-types/contracts/MarginlyFactory';
import { createMarginlyFactory } from './shared/fixtures';
import snapshotGasCost from 'optifat-snapshot-gas-cost';
import { MarginlyPool } from '../typechain-types';
import { ethers } from 'hardhat';
import { PositionType, ZeroAddress } from './shared/utils';

describe('MarginlyFactory', () => {
  function getPoolParams() {
    const params: MarginlyParamsStruct = {
      interestRate: 54000, //5,4 %
      fee: 10000, //1%
      maxLeverage: 20,
      swapFee: 1000, // 0.1%
      mcSlippage: 50000, //5%
      positionMinAmount: 1, // 1 WEI
      quoteLimit: 1_000_000_000_000,
    };

    return {
      params,
      defaultSwapCallData: 0,
    };
  }

  it('should create pool', async () => {
    const { factory, uniswapPoolInfo, priceOracle } = await loadFixture(createMarginlyFactory);
    const quoteToken = uniswapPoolInfo.token0;
    const baseToken = uniswapPoolInfo.token1;
    const { params, defaultSwapCallData } = getPoolParams();

    const poolAddress = await factory.createPool.staticCall(
      quoteToken,
      baseToken,
      priceOracle,
      defaultSwapCallData,
      params
    );
    await snapshotGasCost(factory.createPool(quoteToken, baseToken, priceOracle, defaultSwapCallData, params));

    const poolFactory = await ethers.getContractFactory('MarginlyPool');
    const pool = poolFactory.attach(poolAddress) as MarginlyPool;

    const techPositionOwner = await factory.techPositionOwner();
    const techPosition = await pool.positions(techPositionOwner);
    expect(techPosition._type).to.be.eq(PositionType.Lend);
  });

  it('should change router address', async () => {
    const { factory } = await loadFixture(createMarginlyFactory);
    const routerAddress = await factory.swapRouter();

    await factory.changeSwapRouter(factory);

    const currentRouterAddress = await factory.swapRouter();
    expect(currentRouterAddress).to.be.not.eq(routerAddress);
    expect(currentRouterAddress).to.be.eq(factory);
  });

  it('should create the same pools', async () => {
    const { factory, uniswapPoolInfo, priceOracle } = await loadFixture(createMarginlyFactory);
    const quoteToken = uniswapPoolInfo.token0;
    const baseToken = uniswapPoolInfo.token1;
    const { params, defaultSwapCallData } = getPoolParams();

    await factory.createPool(quoteToken, baseToken, priceOracle, defaultSwapCallData, params);

    await factory.createPool(quoteToken, baseToken, priceOracle, defaultSwapCallData, params);
  });

  it('should raise error when trying to renounce ownership', async () => {
    const { factory } = await loadFixture(createMarginlyFactory);

    await expect(factory.renounceOwnership()).to.be.revertedWithCustomError(factory, 'Forbidden');
  });

  it('should raise error when trying to deploy factory with wrong arguments', async () => {
    const factoryFactory = await ethers.getContractFactory('MarginlyFactory');
    const nonZeroAddress = '0x0000000000000000000000000000000000000001';

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const constructorArgs: [any, any, any, any, any] = [
      nonZeroAddress,
      nonZeroAddress,
      nonZeroAddress,
      nonZeroAddress,
      nonZeroAddress,
    ];

    for (let i = 0; i < constructorArgs.length; i++) {
      constructorArgs[i] = ZeroAddress;
      await expect(factoryFactory.deploy.call(factoryFactory, ...constructorArgs)).to.be.revertedWithCustomError(
        factoryFactory,
        'WrongValue'
      );
      constructorArgs[i] = nonZeroAddress;
    }
  });
});
