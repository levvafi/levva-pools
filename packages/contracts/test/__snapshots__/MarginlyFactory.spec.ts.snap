// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarginlyFactory should create pool 1`] = `
Object {
  "calldataByteLength": 356,
  "gasUsed": 396354,
}
`;
